### PR TITLE
feat: internalChecksFilter

### LINKED ISSUE #7603: Lazy vs Eager internal status checks
**What would you like Renovate to be able to do?**

Better handling of `stabilityDays`. e.g. today's behavior is:

- Assume current dependency version is `1.0.0`
- Assume `stabilityDays=3`
- New version `1.0.1` is released
- After 3 days, branch and PR is created, `renovate/stability-days` is set to green
- A short time after, `1.0.2` is released
- Renovate updates the branch to `1.0.2` and sets `renovate/stability-days` to pending

Most people would prefer that the branch/PR are not updated until `1.0.2` passes stability checks too.

**Did you already have any implementation ideas?**

One possibility is to move `stabilityDays` logic into our "filter" stage, e.g. to filter out versions not exceeding stabilityDays the same way that we today filter out unstable versions.

However the downside of that is lack of visibility. The user would initially have no visibility into `1.0.1` in e.g. Dependency Dashboard, and no way to "force it open" like they can today. Similarly, `1.0.2` would be invisible too.

When `1.0.1` is released, it would be the only one in the non-major lookup "bucket". When `1.0.2` is released, both of them would be in it. We ideally wouldn't want users to need to change their *config* just to temporarily get `1.0.1` or `1.0.2`, for instance.

Possibility:
- Similar to today, when `1.0.1` is released, it is visible as a pending status checks branch/PR and can be forced open that way
- [New] Once `1.0.2` is released, the PR body can be updated to indicate that the newer version is available but behind held back due to status checks. The user could tick a checkbox which enables pending releases *for that current PR only*.

**Are there any workarounds or alternative ideas you've tried to avoid needing this feature?**

Remaining with the current approach.

**Is this a feature you'd be interested in implementing yourself?**

Yes, but need help on ideas

### PR DESCRIPTION
**Pull request by rarkins**
## Changes:

Adds new option internalChecksFilter, allowing for more control over stabilityDays filtering.

## Context:

Closes #7603

I've chosen to default behavior to be backwards compatible for now so that we can test it ourselves, and I'd like to change default to `strict` once we're confident there's little chance of edge cases we've missed.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added unit tests, or
- [ ] No new tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
Does not have full coverage yet - I wanted feedback on the approach first

File: docs/usage/configuration-options.md [hunk 1]
```diff
@@ -1028,6 +1028,18 @@ If you wish for Renovate to process only select paths in the repository, use `in
 Alternatively, if you need to just _exclude_ certain paths in the repository then consider `ignorePaths` instead.
 If you are more interested in including only certain package managers (e.g. `npm`), then consider `enabledManagers` instead.
 
+## internalChecksFilter
+
+This setting controls whether Renovate controls when and how filtering of internal checks are performed, particularly when multiple versions of the same update type are available.
+Currently this applies to the stabilityDays check only.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1034]:
```suggestion
Currently this applies to the `stabilityDays` check only.
```

File: docs/usage/configuration-options.md [hunk 2]
```diff
@@ -1028,6 +1028,18 @@ If you wish for Renovate to process only select paths in the repository, use `in
...
+
+- `none`: No filtering will be performed, and the highest release will be used regardless of whether it's pending or not
+- `strict`: All pending releases will be filtered if they are pending. PRs will be skipped unless a non-pending version is available.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1037]:
```suggestion
- `strict`: All pending releases will be filtered if they are pending. PRs will be skipped unless a non-pending version is available
```

File: docs/usage/configuration-options.md [hunk 3]
```diff
@@ -1028,6 +1028,18 @@ If you wish for Renovate to process only select paths in the repository, use `in
...
+- `flexible`: Similar to strict, but in the case where all versions are pending then a PR will be created with the highest pending version.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1038]:
```suggestion
- `flexible`: Similar to strict, but in the case where all versions are pending then a PR will be created with the highest pending version
```

**Review by viceice — COMMENTED** [review]
Looks good so far, little tricky to review because of the missing coverage messages 🙃

File: docs/usage/configuration-options.md [hunk 4]
```diff
@@ -1028,6 +1028,18 @@ If you wish for Renovate to process only select paths in the repository, use `in
 Alternatively, if you need to just _exclude_ certain paths in the repository then consider `ignorePaths` instead.
 If you are more interested in including only certain package managers (e.g. `npm`), then consider `enabledManagers` instead.
 
+## internalChecksFilter
+
+This setting controls whether Renovate controls when and how filtering of internal checks are performed, particularly when multiple versions of the same update type are available.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1033]:
```suggestion
This setting determines whether Renovate controls when and how filtering of internal checks are performed, particularly when multiple versions of the same update type are available.
```

I don't like that we use the word "controls" twice in the same sentence. I don't know if a word like "determines" is better?

File: docs/usage/configuration-options.md [hunk 5]
```diff
@@ -1028,6 +1028,18 @@ If you wish for Renovate to process only select paths in the repository, use `in
...
+Currently this applies to the stabilityDays check only.
+
+- `none`: No filtering will be performed, and the highest release will be used regardless of whether it's pending or not
+- `strict`: All pending releases will be filtered if they are pending. PRs will be skipped unless a non-pending version is available.
+- `flexible`: Similar to strict, but in the case where all versions are pending then a PR will be created with the highest pending version.
+
+The `flexible` mode can reduce in "flapping" of Pull Requests, where e.g. a pending PR with version `1.0.3` is first released but then downgraded to `1.0.2` once it passes `stabilityDays`.
+It is recommended that you use `strict` combined with `dependencyDashboard` so that you have visibility into suppressed PRs.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1041]:
```suggestion
We recommend that you use the `strict` mode, and enable the `dependencyDashboard` so that you have visibility into suppressed PRs.
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: docs/usage/configuration-options.md [hunk 6]
```diff
@@ -1028,6 +1028,18 @@ If you wish for Renovate to process only select paths in the repository, use `in
...
+## internalChecksFilter
+
+This setting controls whether Renovate controls when and how filtering of internal checks are performed, particularly when multiple versions of the same update type are available.
+Currently this applies to the stabilityDays check only.
+
+- `none`: No filtering will be performed, and the highest release will be used regardless of whether it's pending or not
+- `strict`: All pending releases will be filtered if they are pending. PRs will be skipped unless a non-pending version is available.
+- `flexible`: Similar to strict, but in the case where all versions are pending then a PR will be created with the highest pending version.
+
+The `flexible` mode can reduce in "flapping" of Pull Requests, where e.g. a pending PR with version `1.0.3` is first released but then downgraded to `1.0.2` once it passes `stabilityDays`.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1040]:
```suggestion
The `flexible` mode can result in "flapping" of Pull Requests, where e.g. a pending PR with version `1.0.3` is first released but then downgraded to `1.0.2` once it passes `stabilityDays`.
```

I think you meant `result` instead of `reduce`. 😉

File: docs/usage/configuration-options.md [hunk 7]
```diff
@@ -1028,6 +1028,18 @@ If you wish for Renovate to process only select paths in the repository, use `in
 Alternatively, if you need to just _exclude_ certain paths in the repository then consider `ignorePaths` instead.
 If you are more interested in including only certain package managers (e.g. `npm`), then consider `enabledManagers` instead.
 
+## internalChecksFilter
+
+This setting determines whether Renovate controls when and how filtering of internal checks are performed, particularly when multiple versions of the same update type are available.
+Currently this applies to the `stabilityDays` check only.
+
+- `none`: No filtering will be performed, and the highest release will be used regardless of whether it's pending or not
+- `strict`: All pending releases will be filtered if they are pending. PRs will be skipped unless a non-pending version is available
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1037]:
```suggestion
- `strict`: All pending releases will be filtered. PRs will be skipped unless a non-pending version is available
```

I think "all pending releases will be filtered" doesn't need the "if they are pending" qualifier?

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/workers/repository/process/lookup/filter-checks.spec.ts [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        versioning,
+        'patch',
+        sortedReleases
+      );
+      expect(res).toMatchSnapshot();
+      expect(res.pendingChecks).toHaveLength(0);
+      expect(res.pendingReleases).toHaveLength(0);
+      expect(res.release.version).toEqual('1.0.4');
+    });
+    it('returns non-pending latest release if internalChecksFilter=flexible and none pass checks', () => {
```
Comment (viceice) [inline lib/workers/repository/process/lookup/filter-checks.spec.ts:62]:
```suggestion

    it('returns non-pending latest release if internalChecksFilter=flexible and none pass checks', () => {
```

Some newlines between tests for better readability?

File: lib/workers/repository/process/lookup/filter-checks.spec.ts [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        versioning,
+        'patch',
+        sortedReleases
+      );
+      expect(res).toMatchSnapshot();
+      expect(res.pendingChecks).toHaveLength(0);
+      expect(res.pendingReleases).toHaveLength(0);
+      expect(res.release.version).toEqual('1.0.4');
+    });
+    it('returns pending latest release if internalChecksFilter=strict and none pass checks', () => {
```
Comment (viceice) [inline lib/workers/repository/process/lookup/filter-checks.spec.ts:76]:
```suggestion

    it('returns pending latest release if internalChecksFilter=strict and none pass checks', () => {
```

File: lib/workers/repository/process/lookup/filter-checks.spec.ts [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+        versioning,
+        'patch',
+        sortedReleases
+      );
+      expect(res).toMatchSnapshot();
+      expect(res.pendingChecks).toHaveLength(1);
+      expect(res.pendingReleases).toHaveLength(0);
+      expect(res.release.version).toEqual('1.0.4');
+    });
+    it('returns non-latest release if internalChecksFilter=strict and some pass checks', () => {
```
Comment (viceice) [inline lib/workers/repository/process/lookup/filter-checks.spec.ts:90]:
```suggestion

    it('returns non-latest release if internalChecksFilter=strict and some pass checks', () => {
```

File: lib/workers/repository/process/lookup/filter-checks.spec.ts [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+        versioning,
+        'patch',
+        sortedReleases
+      );
+      expect(res).toMatchSnapshot();
+      expect(res.pendingChecks).toHaveLength(0);
+      expect(res.pendingReleases).toHaveLength(2);
+      expect(res.release.version).toEqual('1.0.2');
+    });
+    it('returns non-latest release if internalChecksFilter=flexible and some pass checks', () => {
```
Comment (viceice) [inline lib/workers/repository/process/lookup/filter-checks.spec.ts:104]:
```suggestion

    it('returns non-latest release if internalChecksFilter=flexible and some pass checks', () => {
```

File: lib/workers/repository/process/lookup/filter-checks.spec.ts [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+        versioning,
+        'patch',
+        sortedReleases
+      );
+      expect(res).toMatchSnapshot();
+      expect(res.pendingChecks).toHaveLength(0);
+      expect(res.pendingReleases).toHaveLength(2);
+      expect(res.release.version).toEqual('1.0.2');
+    });
+    it('picks up stabilityDays settings from hostRules', () => {
```
Comment (viceice) [inline lib/workers/repository/process/lookup/filter-checks.spec.ts:118]:
```suggestion

    it('picks up stabilityDays settings from hostRules', () => {
```

File: lib/workers/repository/process/lookup/filter-checks.spec.ts [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+        versioning,
+        'patch',
+        sortedReleases
+      );
+      expect(res).toMatchSnapshot();
+      expect(res.pendingChecks).toHaveLength(0);
+      expect(res.pendingReleases).toHaveLength(0);
+      expect(res.release.version).toEqual('1.0.4');
+    });
+    it('picks up stabilityDays settings from updateType', () => {
```
Comment (viceice) [inline lib/workers/repository/process/lookup/filter-checks.spec.ts:133]:
```suggestion

    it('picks up stabilityDays settings from updateType', () => {
```

File: lib/workers/branch/index.spec.ts [hunk 14]
```diff
@@ -324,6 +324,17 @@ describe(getName(), () => {
       commit.commitFilesToBranch.mockResolvedValueOnce(null);
       expect(await branchWorker.processBranch(config)).toMatchSnapshot();
     });
+    it('returns if pending checks', async () => {
```
Comment (viceice) [inline lib/workers/branch/index.spec.ts:327]:
```suggestion

    it('returns if pending checks', async () => {
```

File: lib/workers/branch/index.spec.ts [hunk 15]
```diff
@@ -324,6 +324,17 @@ describe(getName(), () => {
...
+        ...updatedPackageFiles,
+      });
+      npmPostExtract.getAdditionalFiles.mockResolvedValueOnce({
+        artifactErrors: [],
+        updatedArtifacts: [],
+      });
+      config.pendingChecks = ['stabilityDays'];
+      expect(await branchWorker.processBranch(config)).toMatchSnapshot();
+    });
     it('returns if branch automerged', async () => {
```
Comment (viceice) [inline lib/workers/branch/index.spec.ts:338]:
```suggestion

    it('returns if branch automerged', async () => {
```

File: lib/workers/repository/process/lookup/filter-checks.ts [hunk 16]
```diff
@@ -0,0 +1,81 @@
+import is from '@sindresorhus/is';
+import { mergeChildConfig } from '../../../../config';
+import { Release } from '../../../../datasource';
+import { logger } from '../../../../logger';
+import { getElapsedDays } from '../../../../util/date';
+import { applyPackageRules } from '../../../../util/package-rules';
+import { VersioningApi } from '../../../../versioning';
+import { LookupUpdateConfig, UpdateResult } from './types';
```
Comment (viceice) [inline lib/workers/repository/process/lookup/filter-checks.ts:8]:
```suggestion
import type { VersioningApi } from '../../../../versioning';
import type { LookupUpdateConfig, UpdateResult } from './types';
```

File: lib/workers/repository/process/lookup/filter-checks.ts [hunk 17]
```diff
@@ -0,0 +1,81 @@
+import is from '@sindresorhus/is';
+import { mergeChildConfig } from '../../../../config';
+import { Release } from '../../../../datasource';
```
Comment (viceice) [inline lib/workers/repository/process/lookup/filter-checks.ts:3]:
```suggestion
import type { Release } from '../../../../datasource';
```

File: lib/workers/repository/process/lookup/filter-checks.spec.ts [hunk 18]
```diff
@@ -0,0 +1,148 @@
+import { getConfig, getName, mocked } from '../../../../../test/util';
+import { Release } from '../../../../datasource';
```
Comment (viceice) [inline lib/workers/repository/process/lookup/filter-checks.spec.ts:2]:
```suggestion
import type { Release } from '../../../../datasource';
```

File: lib/workers/repository/process/lookup/filter-checks.spec.ts [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+import { clone } from '../../../../util/clone';
+import * as _dateUtil from '../../../../util/date';
+import * as allVersioning from '../../../../versioning';
+import { filterInternalChecks } from './filter-checks';
+import { LookupUpdateConfig, UpdateResult } from './types';
```
Comment (viceice) [inline lib/workers/repository/process/lookup/filter-checks.spec.ts:7]:
```suggestion
import type { LookupUpdateConfig, UpdateResult } from './types';
```

**Review by viceice — COMMENTED** [review]
Looks good so far, some small code style issues

File: docs/usage/configuration-options.md [hunk 20]
```diff
@@ -1040,6 +1040,18 @@ If you wish for Renovate to process only select paths in the repository, use `in
...
+
+This setting determines whether Renovate controls when and how filtering of internal checks are performed, particularly when multiple versions of the same update type are available.
+Currently this applies to the `stabilityDays` check only.
+
+- `none`: No filtering will be performed, and the highest release will be used regardless of whether it's pending or not
+- `strict`: All pending releases will be filtered. PRs will be skipped unless a non-pending version is available
+- `flexible`: Similar to strict, but in the case where all versions are pending then a PR will be created with the highest pending version
+
+The `flexible` mode can result in "flapping" of Pull Requests, where e.g. a pending PR with version `1.0.3` is first released but then downgraded to `1.0.2` once it passes `stabilityDays`.
+We recommend that you use the `strict` mode, and enable the `dependencyDashboard` so that you have visibility into suppressed PRs.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1053]: Do we need to mention what the default setting is for `internalChecksFilter`?

It's not clear from reading the current docs what the default behavior is if you only turn on `stabilityDays` and do not set `internalChecksFilter` explicitly yourself.

Comment (viceice) [reply docs/usage/configuration-options.md:1053]: I think the docs generator will add the defaults from the `definitions.ts`

File: lib/workers/repository/process/lookup/index.spec.ts [hunk 21]
```diff
@@ -656,6 +656,66 @@ describe(getName(), () => {
       });
       expect((await lookup.lookupUpdates(config)).updates).toMatchSnapshot();
     });
+    it('should return pendingChecks', async () => {
```
Comment (viceice) [inline lib/workers/repository/process/lookup/index.spec.ts:659]:
```suggestion

    it('should return pendingChecks', async () => {
```

😉

File: lib/workers/repository/process/lookup/index.spec.ts [hunk 22]
```diff
@@ -656,6 +656,66 @@ describe(getName(), () => {
...
+      config.currentValue = '1.4.4';
+      config.depName = 'some/action';
+      config.datasource = datasourceGithubReleases.id;
+      config.stabilityDays = 14;
+      config.internalChecksFilter = 'strict';
+      const yesterday = new Date();
+      yesterday.setDate(yesterday.getDate() - 1);
```
Comment (viceice) [inline lib/workers/repository/process/lookup/index.spec.ts:666]: Better use luxon DateTime?

File: lib/workers/repository/process/lookup/index.spec.ts [hunk 23]
```diff
@@ -656,6 +656,66 @@ describe(getName(), () => {
...
+            releaseTimestamp: yesterday.toISOString(),
+          },
+        ],
+      });
+      const res = await lookup.lookupUpdates(config);
+      expect(res.updates).toHaveLength(1);
+      expect(res.updates[0].newVersion).toEqual('1.4.6');
+      expect(res.updates[0].pendingChecks).toHaveLength(1);
+    });
+    it('should return pendingVersions', async () => {
```
Comment (viceice) [inline lib/workers/repository/process/lookup/index.spec.ts:689]:
```suggestion

    it('should return pendingVersions', async () => {
```

File: lib/workers/repository/process/lookup/index.spec.ts [hunk 24]
```diff
@@ -656,6 +656,66 @@ describe(getName(), () => {
...
+            version: '1.4.6',
+            releaseTimestamp: yesterday.toISOString(),
+          },
+        ],
+      });
+      const res = await lookup.lookupUpdates(config);
+      expect(res.updates).toHaveLength(1);
+      expect(res.updates[0].newVersion).toEqual('1.4.5');
+      expect(res.updates[0].pendingVersions).toHaveLength(1);
+    });
```
Comment (viceice) [inline lib/workers/repository/process/lookup/index.spec.ts:718]:
```suggestion
    });

```

File: lib/workers/repository/process/lookup/index.ts [hunk 25]
```diff
@@ -201,11 +203,21 @@ export async function lookupUpdates(
...
       );
-      const release = sortedReleases.pop();
+      const { release, pendingChecks, pendingReleases } = filterInternalChecks(
+        depResultConfig,
+        versioning,
+        bucket,
+        sortedReleases
+      );
+      // istanbul ignore next
+      if (!release) {
```
Comment (viceice) [inline lib/workers/repository/process/lookup/index.ts:218]: Can this happen? Then we should better test it.

File: lib/workers/repository/process/lookup/index.spec.ts (hunk 22 above)
Comment (rarkins) [reply lib/workers/repository/process/lookup/index.spec.ts:666]: I think it's simple enough as-is, and just for tests

File: lib/workers/repository/process/lookup/index.ts (hunk 25 above)
Comment (rarkins) [reply lib/workers/repository/process/lookup/index.ts:218]: I couldn't work out a way to reproduce it in tests, but I want to keep it here just in case.

**Review by viceice — APPROVED** [review]
